provider_filetree.go

```diff
@@ -1,42 +1,124 @@
 package configstore
 
 import (
-	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"unicode"
 	"unicode/utf8"
+
+	"github.com/fsnotify/fsnotify"
 )
 
 func fileTreeProvider(s *Store, dirname string) {
+	fileTree(s, dirname, false)
+}
+
+func fileTreeRefreshProvider(s *Store, dirname string) {
+	fileTree(s, dirname, true)
+}
+
+func fileTree(s *Store, dirname string, refresh bool) {
 	if dirname == "" {
 		return
 	}
 
-	providername := fmt.Sprintf("filetree:%s", dirname)
+	providername := buildProviderName("filetree", refresh, dirname)
+	inmem := inMemoryProvider(s, providername)
 
-	files, err := ioutil.ReadDir(dirname)
+	items, err := loadItems(dirname)
 	if err != nil {
 		errorProvider(s, providername, err)
-		return
+	} else {
+		inmem.mut.Lock()
+		inmem.items = items
+		inmem.mut.Unlock()
+	}
+
+	if refresh {
```

Comment:
I'd suggest to do the opposite:

```go
if !refresh {
  return
}
```

This way you avoid scoping the whole subsequent block and, imo, improve readability.